modules/client/src/main/java/org/apache/ignite/internal/jdbc/JdbcConnection.java

```diff
@@ -0,0 +1,749 @@
+import java.util.concurrent.Executor;
+import org.apache.ignite.client.IgniteClient;
+import org.apache.ignite.client.proto.query.JdbcQueryEventHandler;
+import org.apache.ignite.client.proto.query.SqlStateCode;
+import org.apache.ignite.internal.client.HostAndPortRange;
+import org.apache.ignite.internal.client.TcpIgniteClient;
+import org.apache.ignite.internal.client.query.JdbcClientQueryEventHandler;
+import org.apache.ignite.schema.SchemaTable;
+import org.jetbrains.annotations.Nullable;
+
+import static java.sql.ResultSet.CLOSE_CURSORS_AT_COMMIT;
+import static java.sql.ResultSet.CONCUR_READ_ONLY;
+import static java.sql.ResultSet.HOLD_CURSORS_OVER_COMMIT;
+import static java.sql.ResultSet.TYPE_FORWARD_ONLY;
+import static org.apache.ignite.client.proto.query.SqlStateCode.CONNECTION_CLOSED;
+
+/**
+ * JDBC connection implementation.
+ */
+public class JdbcConnection implements Connection {
+    /** Network timeout permission. */
+    private static final String SET_NETWORK_TIMEOUT_PERM = "setNetworkTimeout";
+
+    /** Statements modification mutex. */
+    private final Object stmtsMux = new Object();
+
+    /** Handler. */
+    private final JdbcQueryEventHandler handler;
+
+    /** Schema name. */
+    private String schema;
+
+    /** Closed flag. */
+    private volatile boolean closed;
+
+    /** Current transaction isolation. */
+    private int txIsolation;
+
+    /** Auto-commit flag. */
+    private boolean autoCommit;
+
+    /** Read-only flag. */
+    private boolean readOnly;
+
+    /** Current transaction holdability. */
+    private int holdability;
+
+    /** Connection properties. */
+    private final ConnectionProperties connProps;
+
+    /** Tracked statements to close on disconnect. */
+    private final Set<JdbcStatement> stmts = Collections.newSetFromMap(new IdentityHashMap<>());
+
+    /** Network timeout. */
+    private int netTimeout;
+
+    /** Query timeout. */
+    private final @Nullable Integer qryTimeout;
+
+    /** Ignite remote client. */
+    private final TcpIgniteClient client;
+
+    /**
+     * Constructor.
+     *
+     * @param handler Handler.
+     * @param props Properties.
+     */
+    public JdbcConnection(JdbcQueryEventHandler handler, ConnectionProperties props) {
+        this.connProps = props;
+        this.handler = handler;
+
+        autoCommit = true;
+
+        netTimeout = connProps.getConnectionTimeout();
+        qryTimeout = connProps.getQueryTimeout();
+
+        holdability = HOLD_CURSORS_OVER_COMMIT;
+
+        schema = SchemaTable.DEFAULT_SCHEMA_NAME;
+
+        client = null;
+    }
+
+    /**
+     * Creates new connection.
+     *
+     * @param props Connection properties.
+     */
+    public JdbcConnection(ConnectionProperties props) {
+        this.connProps = props;
+        autoCommit = true;
+
+        var objects = Arrays.stream(props.getAddresses()).map(HostAndPortRange::toString).toArray(String[]::new);
+
+        client = ((TcpIgniteClient)IgniteClient
```

Comment:
Why netTimeout isn't used to configure client?
Should we add other client's options to the JDBC properties?

Reply:
connectTimeout and other functions in TcpIgniteClient#Builder was introduced after I made this changes. 
I fixed it and added retryLimit, reconnectThrottlingPeriod and reconnectThrottlingRetries properties as well.